Add ROCm libraries benchmark tests

lajagapp opened the pull request:
## Summary

Comprehensive benchmark testing framework for ROCm libraries with automated system detection, results collection with API integration, performance tracking with LKG comparison, and parallel CI execution.

## Motivation

Implement automated benchmark testing for ROCm libraries (ROCfft, ROCrand, ROCsolver, hipBLASLt) with:
- **Standardized execution framework** via base class pattern
- **Automated system detection** (GPU, OS, ROCm version)
- **Results tracking** with API upload and LKG performance comparison
- **CI integration** with parallel execution for faster nightly builds
- **Extensible architecture** for adding new benchmarks easily

## Technical Details

### Framework Architecture

**Modular benchmark package** (`build_tools/github_actions/benchmarks/`):

1. **Scripts** (5 files)
   - `benchmark_base.py` - Base class with common functionality
   - `test_hipblaslt_benchmark.py` - GEMM operations benchmarks
   - `test_rocfft_benchmark.py` - FFT transformation benchmarks
   - `test_rocrand_benchmark.py` - Random number generation benchmarks
   - `test_rocsolver_benchmark.py` - Linear algebra benchmarks

2. **Configuration** (3 files)
   - `config.yml` - Framework configuration (API endpoints, logging, system info)
   - `hipblaslt.json` - Test case definitions for hipBLASLt
   - `rocfft.json` - Test case definitions for ROCfft
   - `benchmark_test_matrix.py` - CI matrix definitions

3. **Utilities** (15+ modules)
   - **Core**: `BenchmarkClient` API, logging (Python stdlib), constants, exceptions
   - **Config**: YAML parser with environment variable expansion, schema validation
   - **System**: Hardware detection, OS/platform info, ROCm version detection
   - **Results**: Local JSON storage, RESTful API client with retry logic, LKG comparison, JSON schema validation

4. **Documentation** (3 files)
   - Framework README with architecture and usage examples
   - Utils module documentation

### CI Integration

**Parallel execution workflow**:
- New `test_benchmarks.yml` workflow for benchmark execution
- Runs in parallel with regular tests in nightly CI
- Uses `IS_BENCHMARK_WORKFLOW=true` flag for matrix selection
- Supports dedicated benchmark GPU runners via `benchmark-runs-on` field

**Modified CI workflows**:
- `ci_linux.yml` - Add parallel `test_linux_benchmarks` job
- `ci_nightly.yml` - Pass `benchmark_runs_on` parameter
- `configure_ci.py` - Add benchmark labels for nightly runs
- `fetch_test_configurations.py` - Add benchmark matrix selection logic
- `amdgpu_family_matrix.py` - Add `benchmark-runs-on` runner field

## Test Results

**Benchmark Execution**
- All 4 benchmark implementations execute successfully
- Proper error handling and logging throughout execution

**CI Integration**
- Workflow YAML syntax verified
- Parallel job structure confirmed
- Matrix generation logic tested

## Future Work

- Add dedicated benchmark GPU runners for consistent performance
- Extend to Windows platform
- Add more ROCm libraries (rccl, rocBLAS)

geomin12 commented on build_tools/github_actions/fetch_test_configurations.py:
```diff
@@ -62,6 +62,15 @@ def _get_script_path(script_name: str) -> str:
+        "fetch_artifact_args": "--blas --tests",
+        "timeout_minutes": 60,
+        "test_script": f"python {_get_script_path('test_hipblaslt_benchmark.py')}",
+        # TODO(lajagapp): Add windows test
```
for add windows tests, can we open a Github issue and link it here? so we can keep track?

geomin12 commented on build_tools/github_actions/test_executable_scripts/test_rocfft_benchmark.py:
```diff
@@ -0,0 +1,302 @@
+    NUM_ITERATIONS = 20  # Number of benchmark iterations
+    
+    # Load benchmark configuration
+    config_file = SCRIPT_DIR.parent / 'configs/benchmarks/rocfft.json'
```
same here regarding path comment

lajagapp replied:
Done. Changed path construction to use pathlib's / operator.

geomin12 commented on build_tools/github_actions/test_executable_scripts/test_hipblaslt_benchmark.py:
i notice that the following 4 benchmark python files share similar functions. is there an opportunity to combine the logic? we love re-usability :) and that would be helpful as we continue to add more benchmark tests

lajagapp replied:
Thank you for pointing that out. Would you prefer we incorporate this refactoring in the current PR or handle it in a follow-up to maintain scope?

geomin12 replied:
I think if it is possible to land cleaner code now, let's do it now!

lajagapp replied:
Done. I’ve refactored the benchmarks by introducing a common base class for shared logic, and updated each benchmark to inherit from it.

geomin12 commented on build_tools/github_actions/utils/logger.py:
we typically use the built in logging function for python, not sure if this is necessary?

geomin12 replied:
this can be removed then! if other files are using the python's built-in logging module

geomin12 commented on build_tools/github_actions/utils/results/results_api.py:
```diff
@@ -0,0 +1,437 @@
+"""API client for submitting test results."""
```
can we clarify this more? where is this data going to?

lajagapp replied:
Agreed! Updated the docstring for clarity:

"""API client for submitting benchmark test results to the configured results tracking API.
Uploads performance metrics, system metadata, and test configurations for regression detection and analysis."""

The API endpoint is configured via environment variables (API_URL) for security.

lajagapp commented on requirements-test.txt:
```diff
@@ -4,3 +4,7 @@ pytest-check==2.5.3
+prettytable>=3.0.0
+requests>=2.28.0
+jsonschema>=4.0.0
+packaging>=21.0
```
Yes, updated.

lajagapp commented on build_tools/github_actions/utils/results/results_api.py:
```diff
@@ -0,0 +1,437 @@
+    Returns:
+        Complete results payload for API submission
+    """
+    # Build BM config
```
Yes, updated.

lajagapp commented on build_tools/github_actions/benchmark_scripts/test_hipblaslt_benchmark.py:
```diff
@@ -0,0 +1,353 @@
+    # Compare with LKG
+    log.info("Comparing results with LKG")
+    final_table = client.compare_results(test_name=BENCHMARK_NAME, table=table)
+    log.info(f"\n{final_table}")
```
Done! Added collapsible benchmark results to the workflow step summary using gha_append_step_summary.

lajagapp commented on build_tools/github_actions/utils/results/results_api.py:
```diff
@@ -0,0 +1,437 @@
+        """
+        self.api_url = api_url.rstrip('/')
+        self.fallback_url = fallback_url.rstrip('/') if fallback_url else None
+        self.api_key = api_key
```
Implemented! Moved all benchmark tests to nightly-only execution with `skip_on_smoke: True` flag. They now skip on PRs and run only on scheduled nightly builds where secrets are available.

lajagapp commented on build_tools/github_actions/fetch_test_configurations.py:
```diff
@@ -62,6 +62,15 @@ def _get_script_path(script_name: str) -> str:
+        # TODO(lajagapp): Add windows test
+        "platform": ["linux"],
+        "total_shards": 1,
+    },
```
As discussed in team, enabled the benchmark tests to run only on nightly testing.
- Benchmarks now run nightly-only with `skip_on_smoke: True`
- Skip on PRs and push to main (smoke tests)
- Run only on scheduled nightly builds (full tests)
- Each benchmark: ~10-60 min, total ~2 hours for all 4 tests

geomin12 commented on build_tools/github_actions/utils/results/results_handler.py:
```diff
@@ -0,0 +1,362 @@
+        except Exception as e:
+            log.error(f"Unexpected error submitting to API: {e}")
+            log.warning("Results not submitted - unexpected error")
+            return False
```
something like `response.raise_for_status()`

lajagapp commented on build_tools/github_actions/utils/exceptions.py:
Agreed. I’ve simplified the API error handling by removing custom exceptions. The client now calls the API and handles the standardized error responses accordingly.

geomin12 commented on build_tools/github_actions/utils/test_client.py:
yes that would be great

geomin12 commented on build_tools/github_actions/utils/README.md:
for these utils files, can you add another folder, where it lives called `benchmark`? this way users can identify that these are used for benchmark testing

lajagapp replied:
Done! All benchmark-related code (scripts, configs, utils, matrix, README) is now consolidated in benchmarks/ directory. This makes it clear that everything under benchmarks/ is part of the benchmark testing framework.

geomin12 commented on build_tools/github_actions/benchmark_scripts/benchmark_base.py:
```diff
@@ -0,0 +1,242 @@
+def run_benchmark_main(benchmark_instance):
+    """Run benchmark with standard error handling and exit codes."""
+    try:
+        sys.exit(benchmark_instance.run())
```
let's make python do the error handling, so no sys.exit! let's try to use `raise Exception`

lajagapp replied:
Replaced sys.exit() with raise Exception as suggested.

geomin12 commented on build_tools/github_actions/benchmark_test_matrix.py:
```diff
@@ -0,0 +1,64 @@
+        "fetch_artifact_args": "--blas --tests",
+        "timeout_minutes": 60,
+        "test_script": f"python {_get_benchmark_script_path('test_hipblaslt_benchmark.py')}",
+        # TODO(lajagapp): Add windows test
```
nit, can we add this here too 

` TODO(lajagapp): Add windows support (https://github.com/ROCm/TheRock/issues/2478)`

geomin12 commented on build_tools/github_actions/benchmark_scripts/benchmark_base.py:
```diff
@@ -0,0 +1,242 @@
+        
+        Must be implemented by child class.
+        """
+        pass
```
since each test has their own method, let's delete these since they won't be shared

geomin12 commented on build_tools/github_actions/benchmark_test_matrix.py:
```diff
@@ -0,0 +1,64 @@
+    "hipblaslt_bench": {
+        "job_name": "hipblaslt_bench",
+        "fetch_artifact_args": "--blas --tests",
+        "timeout_minutes": 60,
```
is this timeout sufficient enough? 60 minutes for the entire 1000 iterations?

lajagapp replied:
Yes, the current timeout of 60 minutes is sufficient. These benchmark tests typically complete within 30–45 minutes, leaving a reasonable buffer for variations.

geomin12 commented on build_tools/github_actions/configs/README.md:
for these benchmark files, can we add another folder called benchmark and have this in there?

the README is specific to benchmarks, but not configs

geomin12 commented on build_tools/github_actions/configure_ci.py:
```diff
@@ -506,6 +508,10 @@ def matrix_generator(
 
+        # Add benchmark labels to selected_test_names so they're included in test labels for nightly runs
+        requested_benchmark_names = list(benchmark_matrix.keys())
+        selected_test_names.extend(filter_known_names(requested_benchmark_names, "benchmark"))
```
just for my understanding, will we build artifacts then run benchmark tests? or are we pulling a known artifact already

In this flow, we will build and run tests. however, the benchmark tests can't share the same runners as normal tests (as we need good machines that are separated)

My idea is, we could create an entirely new workflow called `benchmark_tests.yml`, where this will use configure_ci.py but pick up different machines and test matrix instead of combining with the CI nightly.

the benchmark test workflow will run nightly, but completely separate as it requires separate machines too

geomin12 replied:
since this requires some design, let's hop on a call. please schedule some time on calendar for us to discuss

lajagapp commented on build_tools/github_actions/benchmarks/scripts/test_rocfft_benchmark.py:
These benchmark tests typically complete within 30–45 minutes.

geomin12 commented on build_tools/github_actions/benchmarks/utils/README.md:
```diff
@@ -0,0 +1,104 @@
+from utils.system import SystemDetector, HardwareDetector, ROCmDetector
+
+# Results handling
+from utils.results import ResultsHandler, ResultsAPI
```
same for these imports

lajagapp replied:
Yes, updated.

lajagapp commented on build_tools/github_actions/benchmarks/utils/README.md:
````diff
@@ -0,0 +1,104 @@
+## Structure
+
+```
+utils/
````
Yes, updated.

geomin12 commented on build_tools/github_actions/benchmarks/benchmark_test_matrix.py:
```diff
@@ -0,0 +1,64 @@
+from pathlib import Path
+
+# Note: these paths are relative to the repository root.
+BENCHMARK_SCRIPT_DIR = Path("./build_tools/github_actions/benchmarks/scripts")
```
you can actually use something like this:

```suggestion
THIS_SCRIPT_DIR = Path(__file__).resolve().parent
```

geomin12 commented on build_tools/github_actions/benchmarks/__init__.py:
is this needed?

lajagapp commented on build_tools/github_actions/fetch_test_configurations.py:
```diff
@@ -232,21 +233,32 @@ def run():
+    # Determine which test matrix to use
+    if is_benchmark_workflow:
+        # For benchmark workflow, use ONLY benchmark_matrix
+        logging.info("Using benchmark_matrix only (IS_BENCHMARK_WORKFLOW=true)")
```
Updated the log message.

geomin12 commented on build_tools/github_actions/fetch_test_configurations.py:
```diff
@@ -232,47 +233,62 @@ def run():
+        # For benchmark workflow, use ONLY benchmark_matrix
+        logging.info("Using benchmark_matrix only (benchmark tests)")
+        selected_matrix = benchmark_matrix.copy()
+        # Benchmarks don't use test_type/test_labels (all have total_shards=1, no filtering)
```
nit: add this comment under # For benchmark workflow, use...

lajagapp commented on build_tools/github_actions/configure_ci.py:
```diff
@@ -502,6 +504,10 @@ def matrix_generator(
 
+        # Add benchmark labels to selected_test_names so they're included in test labels for nightly runs
+        requested_benchmark_names = list(benchmark_matrix.keys())
+        selected_test_names.extend(filter_known_names(requested_benchmark_names, "benchmark"))
```
Yes, these are already inside the `if is_schedule:` block (line 497). They only run for nightly/scheduled workflows, not for PR or push events.

lajagapp commented:
Yes, you're right.

The benchmark code isn't broken—it executes correctly and produces valid results. The issue is performance variability on shared CI machines causing false-positive failures when comparing against LKG scores from isolated runs.

I've implemented the xfail pattern for all benchmark tests.

- Added `expect_failure: true` to all benchmarks in `benchmark_test_matrix.py`
- Updated `test_component.yml` to use `continue-on-error` for components with `expect_failure`
- Job names now show "(xfail)" indicator for visibility

**TODO:**
```
# TODO: Remove xfail once dedicated performance servers are added in "benchmark-runs-on"
"expect_failure": True,
```

Once dedicated benchmark runners are configured in the GPU matrix (`benchmark-runs-on`), we'll remove the xfail flags.

lajagapp commented:
Verified this xfail change with nightly CI workflow - [https://github.com/ROCm/TheRock/actions/runs/20378063042](https://github.com/ROCm/TheRock/actions/runs/20378063042).

lajagapp commented:
> looks good for the initial check in.
> 
> I think there's still a ton of bloat which can be removed as we continue to iterate on this and get benchmarks at a good spot! Especially once we get benchmarks uploading properly and fully working, we can analyze and see what can be removed
> 
> But thanks for this work!

Thanks for the review! Totally agree - let's get benchmarks working end-to-end first, then we can do a cleanup pass based on what we actually need.

lajagapp commented on .github/workflows/test_component.yml:
```diff
@@ -49,6 +53,11 @@ jobs:
+      # NOTE: These secrets are only required for benchmark results submission in nightly CI runs.
+      # For PR/push workflows, secret retrieval will fail gracefully and benchmarks will skip API submission.
+      BENCHMARK_DB_URL: ${{ secrets.BENCHMARK_DB_URL }}
+      BENCHMARK_DB_FALLBACK_URL: ${{ secrets.BENCHMARK_DB_FALLBACK_URL }}
```
Thanks for the heads up! Just to confirm - are you suggesting we use a GitHub App to generate short-lived tokens for authenticating to the benchmark database API, instead of storing long-lived credentials as secrets? If so, I'll track this as a follow-up improvement once benchmarks are stable!
